flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorSubtaskState.java

```diff
@@ -64,81 +61,64 @@
 	/**
 	 * Snapshot from the {@link org.apache.flink.runtime.state.OperatorStateBackend}.
 	 */
-	@Nonnull
 	private final StateObjectCollection<OperatorStateHandle> managedOperatorState;
 
 	/**
 	 * Snapshot written using {@link org.apache.flink.runtime.state.OperatorStateCheckpointOutputStream}.
 	 */
-	@Nonnull
 	private final StateObjectCollection<OperatorStateHandle> rawOperatorState;
 
 	/**
 	 * Snapshot from {@link org.apache.flink.runtime.state.KeyedStateBackend}.
 	 */
-	@Nonnull
 	private final StateObjectCollection<KeyedStateHandle> managedKeyedState;
 
 	/**
 	 * Snapshot written using {@link org.apache.flink.runtime.state.KeyedStateCheckpointOutputStream}.
 	 */
-	@Nonnull
 	private final StateObjectCollection<KeyedStateHandle> rawKeyedState;
 
-	@Nonnull
 	private final StateObjectCollection<InputChannelStateHandle> inputChannelState;
 
-	@Nonnull
 	private final StateObjectCollection<ResultSubpartitionStateHandle> resultSubpartitionState;
 
 	/**
-	 * The state size. This is also part of the deserialized state handle.
-	 * We store it here in order to not deserialize the state handle when
-	 * gathering stats.
+	 * The subpartitions mappings per partition set when the output operator for a partition was rescaled. The key is
+	 * the partition id and the value contains all subtask indexes of the output operator before rescaling.
 	 */
-	private final long stateSize;
+	private final InflightDataRescalingDescriptor inputRescalingDescriptor;
```

Comment:
I think this field needs to be (de)serialized by `MetadataSerializer` as other fields are.

(ditto `outputRescalingDescriptor`)

Reply:
Hm in the serialized form it would always be empty, so I'm not sure what the advantage is (it can only be filled when you know the post-scale parallelism, which you wouldn't know in a checkpoint).

Reply:
Yes, you are right (nit: maybe comment that it's only computed upon assignment?)

Reply:
Added a comment.